xray/utils/sarifutils.go

```diff
@@ -1,111 +1,344 @@
+				continue
+			}
+			// Result exists at source, compare locations info
+			for _, matchingSourceResult := range matchingSourceResults {
+				if len(targetRuleResult.Locations) > len(matchingSourceResult.Locations) &&
+					len(targetRuleResult.CodeFlows) > len(matchingSourceResult.CodeFlows) {
+					runWithNewOnly.AddResult(targetRuleResult)
+					runWithNewOnly.Tool.Driver.AddRule(targetRule)
+				}
+			}
+		}
 	}
 	return
 }
 
-func extractThreadFlows(threadFlows []*sarif.ThreadFlow, workingDir string) (flows []*[]SourceCodeLocation) {
-	for _, threadFlow := range threadFlows {
-		if threadFlow == nil || len(threadFlow.Locations) == 0 {
+// Calculate new information that exists at the result and not at the source
+func GetDiffFromResult(result *sarif.Result, source *sarif.Result) *sarif.Result {
+	newLocations := datastructures.MakeSet[*sarif.Location]()
+	newCodeFlows := []*sarif.CodeFlow{}
+	for _, targetLocation := range result.Locations {
+		if !IsLocationInResult(targetLocation, source) {
+			newLocations.Add(targetLocation)
+			newCodeFlows = append(newCodeFlows, GetLocationRelatedCodeFlowsFromResult(targetLocation, result)...)
 			continue
 		}
-		flow := extractStackTraceLocations(threadFlow.Locations, workingDir)
-		if len(flow) > 0 {
-			flows = append(flows, &flow)
+		// Location in result, compare related code flows
+		for _, targetCodeFlow := range GetLocationRelatedCodeFlowsFromResult(targetLocation, result) {
+			for _, sourceCodeFlow := range GetLocationRelatedCodeFlowsFromResult(targetLocation, source) {
+				if !IsSameCodeFlow(targetCodeFlow, sourceCodeFlow) {
+					// Code flow does not exists at source, add it and it's related location
+					newLocations.Add(targetLocation)
+					newCodeFlows = append(newCodeFlows, targetCodeFlow)
+				}
+			}
+		}
+	}
+	// Create the result only with new information
+	return sarif.NewRuleResult(*result.RuleID).
+		WithKind(*result.Kind).
+		WithMessage(&result.Message).
+		WithLevel(*result.Level).
+		WithLocations(newLocations.ToSlice()).
+		WithCodeFlows(newCodeFlows)
+}
+
+func FilterResultsByRuleIdAndMsgText(source []*sarif.Result, ruleId, msgText string) (results []*sarif.Result) {
+	for _, result := range source {
+		if ruleId == *result.RuleID && msgText == GetResultMsgText(result) {
+			results = append(results, result)
 		}
 	}
 	return
 }
 
-func extractStackTraceLocations(locations []*sarif.ThreadFlowLocation, workingDir string) (flow []SourceCodeLocation) {
-	for _, location := range locations {
-		if location == nil {
-			continue
+func GetLocationRelatedCodeFlowsFromResult(location *sarif.Location, result *sarif.Result) (codeFlows []*sarif.CodeFlow) {
+	for _, codeFlow := range result.CodeFlows {
+		for _, stackTrace := range codeFlow.ThreadFlows {
+			// The threadFlow is reverse stack trace.
+			// The last location is the location that it relates to.
+			if IsSameLocation(location, stackTrace.Locations[len(stackTrace.Locations)-1].Location) {
```

Comment:
len(stackTrace.Locations) can be 0?

Reply:
If codeflow exists it will have at least one thread flow